src/particles/elements/mixin/alignment.H

```diff
@@ -0,0 +1,146 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
+ *           Berkeley National Laboratory (subject to receipt of any required
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_ELEMENTS_MIXIN_ALIGNMENT_H
+#define IMPACTX_ELEMENTS_MIXIN_ALIGNMENT_H
+
+#include "particles/ImpactXParticleContainer.H"
+
+#include <ablastr/constant.H>
+
+#include <AMReX_Math.H>
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+
+
+namespace impactx::elements
+{
+    /** This is a helper class for lattice elements with horizontal/vertical alignment errors
+     */
+    struct Alignment
+    {
+        static constexpr amrex::ParticleReal degree2rad = ablastr::constant::math::pi / 180.0;
+
+        /** A finite-length element
+         *
+         * @param dx horizontal translation error in m
+         * @param dy vertical translation error in m
+         * @param rotation_degree transversal rotation error around the z axis (in the x-y plane) [degrees]
+         */
+        Alignment (
+            amrex::ParticleReal dx,
+            amrex::ParticleReal dy,
+            amrex::ParticleReal rotation_degree
+        )
+        : m_dx(dx), m_dy(dy), m_rotation(rotation_degree * degree2rad)
+        {
+        }
+
+        Alignment () = default;
+        Alignment (Alignment const &) = default;
+        Alignment& operator= (Alignment const &) = default;
+        Alignment (Alignment&&) = default;
+        Alignment& operator= (Alignment&& rhs) = default;
+
+        ~Alignment () = default;
+
+        /** Shift the particle into the alignment error frame
+         *
+         * @param[inout] x horizontal position relative to reference particle
+         * @param[inout] y vertical position relative to reference particle
+         * @param[inout] px horizontal momentum relative to reference particle
+         * @param[inout] py vertical momentum relative to reference particle
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void shift_in (
+            amrex::ParticleReal & AMREX_RESTRICT x,
+            amrex::ParticleReal & AMREX_RESTRICT y,
+            amrex::ParticleReal & AMREX_RESTRICT px,
+            amrex::ParticleReal & AMREX_RESTRICT py
+        ) const
+        {
+            auto const [sin_rotation, cos_rotation] = amrex::Math::sincos(m_rotation);
+
+            // position
+            x -= m_dx;
```

Comment:
```suggestion
            amrex::ParticleReal xc = x - m_dx;
```

Reply:
Rename intermediate (x,y) to (xc,yc) to avoid mixing new and old variable values below.

Reply:
Oh right, that's a bug.

What does the `c` stand for?

Reply:
> Just to clarify: this is purely renaming, right?

Ah no, it's a bug. I cannot rotate x already and then use the rotated x'' to rotate y.